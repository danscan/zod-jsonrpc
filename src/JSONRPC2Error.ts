/**
 * Represents a JSON-RPC 2.0 error.
 */
export class JSONRPC2Error extends Error {
  /** Invalid JSON was received by the server. An error occurred on the server while parsing the JSON text. */
  static ParseError = new JSONRPC2Error(-32700, 'Parse error');

  /** The JSON sent is not a valid Request object. */
  static InvalidRequest = new JSONRPC2Error(-32600, 'Invalid Request');
  static InvalidRequestWithData = (data: unknown) => new JSONRPC2Error(-32600, 'Invalid Request', data);

  /** The method does not exist / is not available. */
  static MethodNotFound = new JSONRPC2Error(-32601, 'Method not found');
  static MethodNotFoundWithData = (data: unknown) => new JSONRPC2Error(-32601, 'Method not found', data);

  /** Invalid method parameter(s). */
  static InvalidParams = new JSONRPC2Error(-32602, 'Invalid params');
  static InvalidParamsWithData = (data: unknown) => new JSONRPC2Error(-32602, 'Invalid params', data);

  /** Internal JSON-RPC error. */
  static InternalError = new JSONRPC2Error(-32603, 'Internal error');
  static InternalErrorWithData = (data: unknown) => new JSONRPC2Error(-32603, 'Internal error', data);

  name = 'JSON-RPC 2.0 Error';

  /** The JSON-RPC 2.0 error code. */
  public readonly code: number;
  /** The error data. */
  public readonly data?: unknown;

  constructor(code: number, message: string, data?: unknown) {
    super(message);
    this.code = code;
    this.data = data;
  }

  toJSON() {
    return {
      code: this.code,
      message: this.message,
      data: this.data,
    };
  }

  toString() {
    const header = `${this.name}: ${this.message} (Code: ${this.code})`;
    const data = this.data ? `\n\nData: ${JSON.stringify(this.data, null, 2)}` : '';
    const stack = this.stack ? `\n\nStack: ${this.stack}` : '';

    return `----\n${header}${data}${stack}\n----`;
  }
}
