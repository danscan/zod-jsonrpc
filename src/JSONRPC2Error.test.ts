import { JSONRPC2Error } from './JSONRPC2Error';
import { describe, it, expect } from 'bun:test';

describe('JSONRPC2Error constructor', () => {
  it('should create an error with a code, message, and data', () => {
    const error = new JSONRPC2Error(-32600, 'test', 'test data');
    expect(error.name).toBe('JSON-RPC 2.0 Error');
    expect(error.code).toBe(-32600);
    expect(error.message).toBe('test');
    expect(error.data).toBe('test data');
  });

  it('should create an error with a code and message', () => {
    const error = new JSONRPC2Error(-32600, 'test');
    expect(error.code).toBe(-32600);
    expect(error.message).toBe('test');
    expect(error.data).toBeUndefined();
  });
});

describe('JSONRPC2Error.ParseError', () => {
  it('should create a standard ParseError', () => {
    const error = JSONRPC2Error.ParseError;
    expect(error instanceof JSONRPC2Error).toBe(true);
    expect(error.code).toBe(-32700);
    expect(error.message).toBe('Parse error');
    expect(error.data).toBeUndefined();
  });
});

describe('JSONRPC2Error.InvalidRequest', () => {
  it('should create a standard InvalidRequest', () => {
    const error = JSONRPC2Error.InvalidRequest;
    expect(error instanceof JSONRPC2Error).toBe(true);
    expect(error.code).toBe(-32600);
    expect(error.message).toBe('Invalid Request');
    expect(error.data).toBeUndefined();
  });
});

describe('JSONRPC2Error.InvalidRequestWithData', () => {
  it('should create a standard InvalidRequest with data', () => {
    const error = JSONRPC2Error.InvalidRequestWithData('test data');
    expect(error instanceof JSONRPC2Error).toBe(true);
    expect(error.code).toBe(-32600);
    expect(error.message).toBe('Invalid Request');
    expect(error.data).toBe('test data');
  });
});

describe('JSONRPC2Error.MethodNotFound', () => {
  it('should create a standard MethodNotFound', () => {
    const error = JSONRPC2Error.MethodNotFound;
    expect(error instanceof JSONRPC2Error).toBe(true);
    expect(error.code).toBe(-32601);
    expect(error.message).toBe('Method not found');
    expect(error.data).toBeUndefined();
  });
});

describe('JSONRPC2Error.InvalidParams', () => {
  it('should create a standard InvalidParams', () => {
    const error = JSONRPC2Error.InvalidParams;
    expect(error instanceof JSONRPC2Error).toBe(true);
    expect(error.code).toBe(-32602);
    expect(error.message).toBe('Invalid params');
    expect(error.data).toBeUndefined();
  });
});

describe('JSONRPC2Error.InvalidParamsWithData', () => {
  it('should create a standard InvalidParams with data', () => {
    const error = JSONRPC2Error.InvalidParamsWithData('test data');
    expect(error instanceof JSONRPC2Error).toBe(true);
    expect(error.code).toBe(-32602);
    expect(error.message).toBe('Invalid params');
    expect(error.data).toBe('test data');
  });
});

describe('JSONRPC2Error.InternalError', () => {
  it('should create a standard InternalError', () => {
    const error = JSONRPC2Error.InternalError;
    expect(error instanceof JSONRPC2Error).toBe(true);
    expect(error.code).toBe(-32603);
    expect(error.message).toBe('Internal error');
    expect(error.data).toBeUndefined();
  });
});

describe('JSONRPC2Error.InternalErrorWithData', () => {
  it('should create a standard InternalError with data', () => {
    const error = JSONRPC2Error.InternalErrorWithData('test data');
    expect(error instanceof JSONRPC2Error).toBe(true);
    expect(error.code).toBe(-32603);
    expect(error.message).toBe('Internal error');
    expect(error.data).toBe('test data');
  });

  it('should include the data in its JSON representation', () => {
    const error = JSONRPC2Error.InternalErrorWithData('test data');
    const json = error.toJSON();
    expect(json).toMatchSnapshot();
    expect(json.data).toBe('test data');
  });
});
