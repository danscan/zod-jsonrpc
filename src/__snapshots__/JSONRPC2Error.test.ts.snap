// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`JSONRPC2Error.InternalErrorWithData should include the data in its JSON representation 1`] = `
{
  "code": -32603,
  "data": "test data",
  "message": "Internal error",
}
`;
